### PR TITLE
Fix an issue with CanJS’s imperative event listener code

### PR DESCRIPTION
**Pull request by chasenlehara**
Thanks to @WebReflection for pointing out the issue: https://twitter.com/WebReflection/status/902534840804941825

### CODE REVIEW

File: libraries/canjs/src/components.js [hunk 1]
```diff
@@ -143,6 +143,8 @@ Component.extend({
     </div>
   `),
   ViewModel: DefineMap.extend({
+    seal: false// this allows webcomponentsjs to define $ on the viewModel
```
Comment (chasenlehara) [inline libraries/canjs/src/components.js]: I’m not really happy with this line. By default, instances of `can-define/map/map` are [sealed](https://canjs.com/doc/can-define/map/map.seal.html), and I get the following error without setting `seal` to `false`:

```
	TypeError: can't define property "$": Object is not extensible (node_modules/@webcomponents/webcomponentsjs/webcomponents-lite.js:71)
	dispatchEvent@node_modules/@webcomponents/webcomponentsjs/webcomponents-lite.js:142:342
	dispatch@webpack:///~/can-util/dom/events/events.js:68:0 <- tests.webpack.js:2086:9
	module.exports@webpack:///~/can-util/dom/dispatch/dispatch.js:18:0 <- tests.webpack.js:3957:9
	handler@webpack:///~/can-util/dom/events/make-mutation-event/make-mutation-event.js:47:0 <- tests.webpack.js:14135:8
	dispatchIfListening/<@webpack:///~/can-util/dom/mutation-observer/document/document.js:25:0 <- tests.webpack.js:14208:3
	dispatchIfListening@webpack:///~/can-util/dom/mutation-observer/document/document.js:24:0 <- tests.webpack.js:14207:2
	setup/mutationData.handler/</<@webpack:///~/can-util/dom/mutation-observer/document/document.js:107:0 <- tests.webpack.js:14290:9
	each@webpack:///~/can-util/js/each/each.js:19:0 <- tests.webpack.js:231:9
	setup/mutationData.handler/<@webpack:///~/can-util/dom/mutation-observer/document/document.js:100:0 <- tests.webpack.js:14283:6
	setup/mutationData.handler@webpack:///~/can-util/dom/mutation-observer/document/document.js:99:0 <- tests.webpack.js:14282:5
	add/observer</<@webpack:///~/can-util/dom/mutation-observer/document/document.js:41:0 <- tests.webpack.js:14224:7
	add/observer<@webpack:///~/can-util/dom/mutation-observer/document/document.js:40:0 <- tests.webpack.js:14223:6
```

I’m having a hard time tracking down what’s causing that error because it doesn’t happen when I debug it in the browser and `webcomponents-lite.js` is minified; looking through the un-minified source, I can see a couple places where `$` is being set, but I’m not sure why.

File: libraries/canjs/src/components.js [hunk 2]
```diff
@@ -151,11 +153,11 @@ Component.extend({
   events: {
     inserted: function() {
       let wc = this.element.querySelector('#wc');
-      wc.addEventListener('camelEvent', this.viewModel.handleEvent.bind(this.viewModel));
+      wc.addEventListener('camelEvent', this.viewModel);
```
Comment (WebReflection) [inline libraries/canjs/src/components.js]: :+1:

File: libraries/canjs/src/components.js [hunk 3]
```diff
@@ -151,11 +153,11 @@ Component.extend({
...
     },
     removed: function() {
       let wc = this.element.querySelector('#wc');
-      wc.removeEventListener('camelEvent', this.viewModel.handleEvent.bind(this.viewModel));
+      wc.removeEventListener('camelEvent', this.viewModel);
```
Comment (WebReflection) [inline libraries/canjs/src/components.js]: :+1:

File: libraries/canjs/src/components.js (hunk 1 above)
Comment (robdodson) [reply libraries/canjs/src/components.js]: @azakus do you know if/where wc-lite defines `$`?

Comment (azakus) [reply libraries/canjs/src/components.js]: GAH, this looks like a closure compiler issue. The compiled version of this code https://github.com/webcomponents/shadydom/blob/master/src/patch-events.js#L311 turns the `__eventWrappers` property of`fnOrObject`, the thing you pass as the second argument to `addEventListener`, into `$`.

I'll make a PR for ShadyDOM, and then a new version for webcomponentsjs.

Comment (azakus) [reply libraries/canjs/src/components.js]: https://github.com/webcomponents/shadydom/pull/174

Comment (azakus) [reply libraries/canjs/src/components.js]: Ok, should be fixed with webcomponentsjs 1.0.10

Comment (chasenlehara) [reply libraries/canjs/src/components.js]: Thank you @azakus! I will test this and update my PR after https://github.com/webcomponents/custom-elements-everywhere/pull/11 gets merged in. 😊

**Comment by robdodson** [comment]
K just merged in 11 so I think we can rebase this one.

File: libraries/canjs/src/components.js (hunk 1 above)
Comment (chasenlehara) [reply libraries/canjs/src/components.js]: @azakus After looking at the `shadydom` change more thoroughly, I understand that it just changed the name of the property that’s set.

The problem I meant to describe in this thread is that _any_ property is set when I call `addEventListener` on a custom element; by default, my `viewModel` object is [sealed](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/seal), so it would be nice if `shadydom` didn't mutate it.

I can work around it just fine, but I thought it was worth bringing up. Thanks for your help! 😊

**Comment by chasenlehara** [comment]
@robdodson Should be good to go now!

File: libraries/canjs/src/components.js [hunk 4]
```diff
@@ -143,21 +151,21 @@ Component.extend({
...
+    handleEvent: function() {
       this.eventHandled = true;
     }
   }),
   events: {
     inserted: function() {
       let wc = this.element.querySelector('#wc');
-      wc.addEventListener('camelEvent',
-        this.viewModel.handleTestEvent.bind(this.viewModel));
+      wc.addEventListener('camelEvent', this.viewModel);
```
Comment (robdodson) [inline libraries/canjs/src/components.js]: this should be changed back to `handleTestEvent`. That's what the other tests are using now.

Comment (chasenlehara) [reply libraries/canjs/src/components.js]: Ah ok, so you explicitly don’t want to use [EventListener’s handleEvent](https://developer.mozilla.org/en-US/docs/Web/API/EventListener#handleEvent())?

Comment (WebReflection) [reply libraries/canjs/src/components.js]: it doesn't matter what they use, `handleEvent` is standard. If your `viewModel` has such method, it's by standards good to be set as listener, which is nothing different than a bind, except you don't need to bind and you can remove it later on whenever you want.

TL;DR it makes no sense to say _don't use handleEvent_ when it's the platform providing such mechanism, right?

Comment (robdodson) [reply libraries/canjs/src/components.js]: I didn't want to use handleEvent because, while it is a standard, it's one that I think most people are unfamiliar with and I didn't want to have to go out of our way to explain it in comments in all of the tests. I'd prefer to stick with the named handler because it's just one less thing to have to deal with.

Comment (WebReflection) [reply libraries/canjs/src/components.js]: if you want to remove that listener later on, it's 2 things that you have to deal.

Readdress that method as bound one and use it for both `addEventListener` and `removeEventListener`.

I don't think it should be the scope of this repository to limit `addEventListener` standard behavior, neither the one that should prevent developers from learning ... quite the opposite, to be honest ... but if it makes your life easier as maintainer of this CanJS part, which I am not sure myself what is this about, then it's OK, your lib, your rules.

Comment (WebReflection) [reply libraries/canjs/src/components.js]: here a link just in case somebody reads and would like to know what is this about:
https://dom.spec.whatwg.org/#callbackdef-eventlistener

**Comment by robdodson** [comment]
Just started reviewing :) <br><br>:octocat: *Sent from [GH](http://nodegh.io).*

**Comment by robdodson** [comment]
merged, thanks for your work on this guys :)
